src/LevelZeroIOGroup.cpp

```diff
@@ -0,0 +1,906 @@
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": TIMESTAMP Signals are for batch use only.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        double result = NAN;
+        auto it = m_signal_available.find(signal_name);
+        if (it != m_signal_available.end()) {
+            result = (it->second.m_signals.at(domain_idx))->read();
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": Handling not defined for " + signal_name,
+                            GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+        return result;
+    }
+
+    // Write to the control immediately, bypassing write_batch()
+    void LevelZeroIOGroup::write_control(const std::string &control_name,
+                                         int domain_type, int domain_idx,
+                                         double setting)
+    {
+        if (!is_valid_control(control_name)) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) + ": "
+                            + control_name + " not valid for LevelZeroIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_type != control_domain_type(control_name)) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) + ": "
+                            + control_name + ": domain_type must be " +
+                            std::to_string(control_domain_type(control_name)),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_idx < 0 ||
+            domain_idx >= m_platform_topo.num_domain(
+                                          control_domain_type(control_name))) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": domain_idx out of range.", GEOPM_ERROR_INVALID,
+                            __FILE__, __LINE__);
+        }
+
+        if (control_name == "LEVELZERO::FREQUENCY_GPU_CONTROL" ||
+            control_name == "FREQUENCY_ACCELERATOR_CONTROL") {
+            m_levelzero_device_pool.frequency_control(domain_type, domain_idx,
+                                                      geopm::LevelZero::M_DOMAIN_COMPUTE,
+                                                      setting / 1e6);
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+                throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                                "Handling not defined for " + control_name,
+                                GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+    }
+
+    // Implemented to allow an IOGroup to save platform settings before starting
+    // to adjust them
+    void LevelZeroIOGroup::save_control(void)
```

Comment:
This is a serious gap in the responsibilities of the IOGroup.  Ideally, all controls that the user has permissions to modify should be saved and restored.